tez-runtime-library/src/main/java/org/apache/tez/runtime/library/common/shuffle/orderedgrouped/ShuffleScheduler.java

```diff
@@ -1015,51 +1019,63 @@ boolean isShuffleHealthy(InputAttemptIdentifier srcAttempt) {
     final float MIN_REQUIRED_PROGRESS_PERCENT = minReqProgressFraction;
     final float MAX_ALLOWED_STALL_TIME_PERCENT = maxStallTimeFraction;
 
-    int doneMaps = numInputs - remainingMaps.get();
+    String errorMsg = null;
```

Comment:
I think instead of creating synchronized block here, it's time to refactor this awful piece of code...we're reporting exceptions and returning with a vague boolean, which is not okay, instead, we should do it in copyFailed: catch all IOExceptions and report them, what about:
```
    //Restart consumer in case shuffle is not healthy
try{
   checkIfShuffleIsHealthy(srcAttempt)
} catch(IOExcepion e){
   exceptionReporter.reportException(ioe);
   return;
}
```

in checkIfShuffleIsHealthy we can do separate things synchronized and don't have to return boolean:
```
void checkIfShuffleIsHealthy(){
   checkIfAbortLimitIsExceeedFor(srcAttempt)
   checkWhateverTheRestOfTheMethodDoes(srcAttempt)
}
```
regarding checkWhateverTheRestOfTheMethodDoes:
1. this is the logic you're about to make synchronized, you can make it I guess (as exception reporting is handled in caller method copyFailed as suggested above
2. find a proper method name for checkWhateverTheRestOfTheMethodDoes, which reflects what it actually does

Reply:
@abstractdog The goal of the patch I submitted was to prevent deadlock, and it was not about simplifying the logic in check???() methods. At the time of submitting the patch, I didn't fully understand the details of the logic in check???() and related methods. Another plan could be to create a new patch for simplifying the logic here.

Reply:
@abstractdog The goal of the patch I submitted was to prevent deadlock, and it was not about simplifying the logic in check???() methods. At the time of submitting the patch, I didn't fully understand the details of the logic in check???() and related methods. Another plan could be to create a new patch for simplifying the logic here.

Reply:
yeah, the original goal has already been achieved I guess
what I really meant was to make a simple refactor in the very same codepath without making the patch bigger, please refer to https://github.com/apache/tez/pull/273